packages/dds/task-manager/src/oldestClientObserver.ts

```diff
@@ -0,0 +1,115 @@
+
+import { AttachState } from "@fluidframework/container-definitions";
+import { IQuorum } from "@fluidframework/protocol-definitions";
+import { IContainerRuntime } from "@fluidframework/container-runtime-definitions";
+import { IOldestClientObserver } from "./interfaces";
+
+/**
+ * The `OldestClientObserver` is a utility inspect if the local client is the oldest amongst connected clients (in
+ * terms of when they connected) and watch for changes.
+ *
+ * It is still experimental and under development.  Please do try it out, but expect breaking changes in the future.
+ *
+ * @remarks
+ * ### Creation
+ *
+ * The `OldestClientObserver` constructor takes an `IContainerRuntime`:
+ *
+ * ```typescript
+ * const oldestClientObserver = new OldestClientObserver(containerRuntime);
+ * ```
+ *
+ * ### Usage
+ *
+ * To check if the local client is the oldest, use the `isOldest()` method.
+ *
+ * ```typescript
+ * if (oldestClientObserver.isOldest()) {
+ *     console.log("I'm the oldest");
+ * } else {
+ *     console.log("Someone else is older");
+ * }
+ * ```
+ *
+ * ### Eventing
+ *
+ * `OldestClientObserver` is an `EventEmitter`, and will emit events when the local client becomes the oldest and when
+ * it is no longer the oldest.
+ *
+ * ```typescript
+ * oldestClientObserver.on("becameOldest", () => {
+ *     console.log("I'm the oldest now");
+ * });
+ *
+ * oldestClientObserver.on("lostOldest", () => {
+ *     console.log("I'm not the oldest anymore");
+ * });
+ * ```
+ */
+export class OldestClientObserver extends EventEmitter implements IOldestClientObserver {
+    private readonly quorum: IQuorum;
+    private currentIsOldest: boolean = false;
+    constructor(private readonly containerRuntime: IContainerRuntime) {
+        super();
+        this.quorum = this.containerRuntime.getQuorum();
+        this.currentIsOldest = this.computeIsOldest();
+        this.quorum.on("addMember", this.updateOldest);
+        this.quorum.on("removeMember", this.updateOldest);
+        containerRuntime.on("connected", this.updateOldest);
+        containerRuntime.on("disconnected", this.updateOldest);
+    }
+
+    public isOldest(): boolean {
+        return this.currentIsOldest;
+    }
+
+    private readonly updateOldest = () => {
+        const oldest = this.computeIsOldest();
+        if (this.currentIsOldest !== oldest) {
+            this.currentIsOldest = oldest;
+            if (oldest) {
+                this.emit("becameOldest");
+            } else {
+                this.emit("lostOldest");
+            }
+        }
+    };
+
+    private computeIsOldest(): boolean {
+        // If the container is detached, we are the only ones that know about it and are the oldest by default.
+        if (this.containerRuntime.attachState === AttachState.Detached) {
+            return true;
+        }
+
+        // Without a clientId we can't verify if we're oldest, assume we are not
+        if (this.containerRuntime.clientId === undefined) {
```

Comment:
I'd move it down and convert to assert - checking for connected is enough

Reply:
Updated in latest.